Update high level constructor

radka-j opened the pull request:
Closes #106 

This PR separates the diameter arguments from the other plankton specific arguments into a separate dictionary. The code is also refactored such that diameters can now be specified as a dictionary which is used to compute an array of values or they can be passed as an array of values by the user. This means that in the case of the user wanting to have a single plankton of diameter 2, the user can just do the following:

```
NP2ZD = construct_size_structured_NPZD(n_phyto=1, phyto_diameters=[2])
```

This is illustrated in tests.

I also declared all dictionaries of default arguments in the global namespace so that the user can access them. This makes updating a single value easy for the user. For example:

```
my_new_args = Agate.Models.Constructors.DEFAULT_PHYTO_ARGS
my_new_args["alpha"] = 0.3953 / day
model_constructed = construct_size_structured_NPZD(
    phyto_args=my_new_args
)
```

nanophyto commented:
> The code is also refactored such that diameters can now be specified as a dictionary which is used to compute an array of values or they can be passed as an array of values by the user. 

This is great, and something I was going to suggest as this will also be useful if n>1.

> I also declared all dictionaries of default arguments in the global namespace so that the user can access them. This makes updating a single value easy for the user. For example:
> 
> ```
> my_new_args = Agate.Models.Constructors.DEFAULT_PHYTO_ARGS
> my_new_args["alpha"] = 0.3953 / day
> model_constructed = construct_size_structured_NPZD(
>     phyto_args=my_new_args
> )
> ```
This looks like a very clean solution. It is not the most intuitive from a UX perspective, but I think it is fit for purpose and we can illustrate the functionality in the docs/example.

radka-j commented:
I plan to revisit this when implementing the `set!()` method, which should allow the user to instantiate a model and then reset specific arguments. I think that what we have here and resetting parameter values might be our only 2 options if we want to also retain the flexibility of the user being able to specify new functions with a completely different set of arguments. For now, I just can't think of a way to have both.

nanophyto commented on src/Models/Constructors.jl:
```diff
@@ -12,6 +12,45 @@ using Oceananigans.Units
 
+DEFAULT_PHYTO_ARGS = Dict(
+    "allometry" => Dict(
+        "maximum_growth_rate" => Dict("a" => 2 / day, "b" => -0.15),
+        "nutrient_half_saturation" => Dict("a" => 0.17, "b" => 0.27),
+    ),
+    "linear_mortality" => 8e-7 / second,
+    "alpha" => 0.1953 / day,
+)
+
+DEFAULT_ZOO_ARGS = Dict(
+    "allometry" => Dict("maximum_predation_rate" => Dict("a" => 30.84 / day, "b" => -0.16)),
+    "linear_mortality" => 8e-7 / second,
+    "holling_half_saturation" => 5.0,
+    "quadratic_mortality" => 1e-6 / second,
+)
+
+DEFAULT_INTERACTION_ARGS = Dict(
+    "P" => Dict(
+        "can_eat" => 0, # bool
+        "can_be_eaten" => 1, # bool
+        "optimum_predator_prey_ratio" => 0,
+        "protection" => 0,
+        "specificity" => 0,
+        "assimilation_efficiency" => 0,
+    ),
+    "Z" => Dict(
+        "can_eat" => 1, # bool
+        "can_be_eaten" => 0, # bool
+        "optimum_predator_prey_ratio" => 10,
+        "protection" => 1,
+        "specificity" => 0.3,
+        "assimilation_efficiency" => 0.32,
+    ),
+)
+
+DEFAULT_BGC_ARGS = Dict(
+    "detritus_remineralization" => 0.1213 / day, "mortality_export_fraction" => 0.5
+)
+
```
Specifying these here might lead to conflicts with other constructors which will have different default arguments.

nanophyto replied:
Instead of moving them they could be renamed to e.g. `SIZE_NPZD_DEFAULT_*` to avoid breaking renaming functionality discussed above?

radka-j replied:
or should we have a constructors director with each their own  module? if we want to have multiple, might be easier to maintain anyways. That way each have their own args and the name of the module will be in the path to access them.

nanophyto reviewed (commented):
This looks great already.

My only other suggestion would be to add a check to ensure e.g. `n_phyto` and `phyto_diameters` are the same length.

radka-j commented:
@nanophyto I moved the constructors to a new director so that each is now its own module and added a check for diameters being the correct length if passed as an array. let me know what you think.

nanophyto reviewed (approved):
LGTM
